src/compiler.rs

```diff
@@ -18,10 +18,21 @@ use crate::Rules;
 ///
 /// add_rules_* functions takes ownership on the compiler, because if a rule fails to compile,
 /// the Compiler is corrupted. See issue [#47](https://github.com/Hugal31/yara-rust/issues/47).
-#[derive(Debug)]
 pub struct Compiler {
     inner: *mut yara_sys::YR_COMPILER,
     _token: InitializationToken,
+    // Safety: this variable should only be edited by `set_include_callback()`
+    // as this needs to be kept alive for include callback to work.
+    include_callback: Option<Box<Box<dyn Fn(&str, Option<&str>, Option<&str>) -> Option<String> >>>,
```

Comment:
Why do you use here double Box? 

`Option<Box<dyn Fn(...) -> Option<String>>`

You can see example: https://github.com/Hugal31/yara-rust/blob/master/src/internals/scan.rs#L176

Reply:
I might need to improve the comment, but it is not quite the same as your example:

In your example:
1. line 176: closure moved into box
2. line 179: the pointer to the box is given to yara for the callback as `user_data`
3. line 184: the pointer can be use by the scan operation
4. on the return the box is destroyed, the `user_data` will no longer be used
=> safety consideration: the box  SHOULD NOT move between the call line 179 and the end of the call line 184

In my case I have the same consideration but with a larger lifetime:
1. closure given to `set_include_callback`
2. closure moved inside a box 1
3. box 1 moved into box 2
4. pointer to box 1 used as `user_data` for the callback in yara C
5. box 2 stored into compiler
6. `set_include_callback()` returns
7. user calls multiple `add_rules_*()` during which the `user_data` can be used
8. the compiler is destroyed, the `user_data` will no longer be used and should be freed
=> safety consideration: the `user_data` MUST remain valid during the all compilation, starting at step 4 where the pointer is take until step 8 it MUST NOT change (not during step 5, not if the compiler is moved, ...).

In a way it might be better to have a `Pin<Box< Box<dyn Fn(...)> >>` which clearly state that the inner box address MUST NOT change.

An other solution, maybe less error prone, would be to have something like:
```rust
pub struct Compiler {
// ...
// The user_data used by the include callback and it's associated free function
include_user_data: Option<(*c_void, fn(*c_void)>
}

impl Drop for Compiler {
    fn drop(&mut self) {
        internals::compiler_destroy(self.inner);
        if let Some((data, free)) = self.include_user_data {
             free(data);
        }
    }
}

impl Compiler {
    pub fn set_include_callback<C>(&mut self, callback: C)
    where
        C: Fn(&str, Option<&str>, Option<&str>) -> Option<String> + 'static,
    {
        use std::os::raw::{c_char, c_void};
        unsafe extern "C" fn include_callback<C>(
            include_name: *const c_char,
            calling_rule_filename: *const c_char,
            calling_rule_namespace: *const c_char,
            user_data: *mut c_void,
        ) -> *const c_char {
            let cb: &mut C = std::mem::transmute(user_data);
            // ...
        }

        fn free_user_data<C>(user_data: *mut c_void) {
            let cb: *mut C = std::mem::transmute(user_data);
            let _ = Box::from_raw(cb);
        }

        let callback = Box::new(callback);
        let user_data = Box::leak(callback) as *mut _ as *mut c_void;
        // ...
        if let Some((data, free)) = self.include_user_data.take() {
             free(data);
        }
        self.include_callback = Some((user_data, free_user_data));
    }
}
```

Reply:
LGTM, thank you!